Gradle QuarkusTest fails with Compilation Failed when using lombok on reload

### Describe the bug

When using the lombok gradle plugin with the `quarkusTest` task , the live reload fails with

```
Compilation Failed:
/src/main/java/org/acme/GreetingResource.java:8: error: package lombok.extern.slf4j does not exist
import lombok.extern.slf4j.Slf4j;
```

This does not happen when running `quarkusDev` , and in continuous testing with `quarkusDev`

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

Reproducer : [https://github.com/Malandril/quarkus-bug-lombok-test](https://github.com/Malandril/quarkus-bug-lombok-test)

run `./gradlew quarkusTest`

Then change anything in the `GreetingResource` class, it wail fail with a compilation error


### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

21

### Quarkus version or git rev

3.29.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

9.1.0

### Additional information

It is caused by #50253 , more precisely the commit 1ed554b6abc831138610d808703d9bf8f0b1f8b0 , reverting it fixes the issue.

--- a/core/deployment/src/main/java/io/quarkus/deployment/dev/QuarkusCompiler.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/dev/QuarkusCompiler.java
@@ -25,6 +25,7 @@ import org.jboss.logging.Logger;
 import io.quarkus.bootstrap.app.CuratedApplication;
 import io.quarkus.bootstrap.app.QuarkusBootstrap;
 import io.quarkus.fs.util.FileSystemProviders;
+import io.quarkus.maven.dependency.DependencyFlags;
 import io.quarkus.maven.dependency.ResolvedDependency;
 import io.quarkus.paths.PathCollection;
 
@@ -67,7 +68,8 @@ public class QuarkusCompiler implements Closeable {
 
         final boolean skipReloadableArtifacts = !application.hasReloadableArtifacts();
 
-        for (ResolvedDependency i : application.getApplicationModel().getRuntimeDependencies()) {
+        for (ResolvedDependency i : application.getApplicationModel()
+                .getDependenciesWithAnyFlag(DependencyFlags.COMPILE_ONLY | DependencyFlags.RUNTIME_CP)) {
             if (skipReloadableArtifacts) {
                 paths.addAll(i.getContentTree().getRoots());
             } else {

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusTest.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusTest.java
@@ -5,7 +5,9 @@ import javax.inject.Inject;
 import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.tasks.Internal;
 
+import io.quarkus.bootstrap.app.QuarkusBootstrap;
 import io.quarkus.deployment.dev.DevModeCommandLineBuilder;
+import io.quarkus.deployment.dev.DevModeContext;
 import io.quarkus.deployment.dev.IsolatedTestModeMain;
 import io.quarkus.gradle.extension.QuarkusPluginExtension;
 import io.quarkus.runtime.LaunchMode;
@@ -22,8 +24,12 @@ public abstract class QuarkusTest extends QuarkusDev {
 
     @Override
     protected void modifyDevModeContext(DevModeCommandLineBuilder builder) {
-        builder.entryPointCustomizer(
-                devModeContext -> devModeContext.setAlternateEntryPoint(IsolatedTestModeMain.class.getName()));
+        builder.entryPointCustomizer(QuarkusTest::configureContinuousTesting);
+    }
+
+    private static void configureContinuousTesting(DevModeContext devModeContext) {
+        devModeContext.setMode(QuarkusBootstrap.Mode.CONTINUOUS_TEST);
+        devModeContext.setAlternateEntryPoint(IsolatedTestModeMain.class.getName());
     }
 
     @Override

--- a/devtools/maven/src/main/java/io/quarkus/maven/TestMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/TestMojo.java
@@ -1,7 +1,5 @@
 package io.quarkus.maven;
 
-import java.util.function.Consumer;
-
 import org.apache.maven.plugins.annotations.LifecyclePhase;
 import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.ResolutionScope;
@@ -25,12 +23,11 @@ public class TestMojo extends DevMojo {
 
     @Override
     protected void modifyDevModeContext(DevModeCommandLineBuilder builder) {
-        builder.entryPointCustomizer(new Consumer<DevModeContext>() {
-            @Override
-            public void accept(DevModeContext devModeContext) {
-                devModeContext.setMode(QuarkusBootstrap.Mode.CONTINUOUS_TEST);
-                devModeContext.setAlternateEntryPoint(IsolatedTestModeMain.class.getName());
-            }
-        });
+        builder.entryPointCustomizer(TestMojo::configureContinuousTesting);
+    }
+
+    private static void configureContinuousTesting(DevModeContext devModeContext) {
+        devModeContext.setMode(QuarkusBootstrap.Mode.CONTINUOUS_TEST);
+        devModeContext.setAlternateEntryPoint(IsolatedTestModeMain.class.getName());
     }
 }

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/QuarkusGradleWrapperTestBase.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/QuarkusGradleWrapperTestBase.java
@@ -332,6 +332,7 @@ public class QuarkusGradleWrapperTestBase extends QuarkusGradleTestBase {
 
                     bw.write(line);
                     bw.newLine();
+                    bw.flush(); // Ensure log is written immediately for real-time reading
 
                     if (forwardToStdOut) {
                         System.out.println(line);

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/continuoustesting/ContinuousTestingClient.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/continuoustesting/ContinuousTestingClient.java
@@ -17,7 +17,7 @@ import io.quarkus.devui.tests.DevUIJsonRPCTest;
 public class ContinuousTestingClient {
     private static final int DEFAULT_PORT = 8080;
 
-    long runtToWaitFor = 1;
+    long runToWaitFor = 1;
     final String host;
 
     protected static String getDefaultHost(int port) {
@@ -40,18 +40,18 @@ public class ContinuousTestingClient {
         try {
             Awaitility.waitAtMost(2, TimeUnit.MINUTES).pollInterval(200, TimeUnit.MILLISECONDS).until(() -> {
                 TestStatus ts = getTestStatus();
-                if (ts.getLastRun() > runtToWaitFor) {
+                if (ts.getLastRun() > runToWaitFor) {
                     throw new RuntimeException(
-                            "Waiting for run " + runtToWaitFor + " but run " + ts.getLastRun() + " has already occurred");
+                            "Waiting for run " + runToWaitFor + " but run " + ts.getLastRun() + " has already occurred");
                 }
-                boolean runComplete = ts.getLastRun() == runtToWaitFor;
+                boolean runComplete = ts.getLastRun() == runToWaitFor;
                 if (runComplete && ts.getRunning() > 0) {
                     //there is a small chance of a race, where changes are picked up twice, due to how filesystems work
                     //this works around it by waiting for the next run
-                    runtToWaitFor = ts.getRunning();
+                    runToWaitFor = ts.getRunning();
                     return false;
                 } else if (runComplete) {
-                    runtToWaitFor++;
+                    runToWaitFor++;
                 }
                 return runComplete;
             });
@@ -60,10 +60,12 @@ public class ContinuousTestingClient {
             TestStatus ts;
             try {
                 ts = getTestStatus();
+            } catch (RuntimeException ex) {
+                throw ex;
             } catch (Exception ex) {
                 throw new RuntimeException(ex);
             }
-            throw new ConditionTimeoutException("Failed to wait for test run " + runtToWaitFor + " " + ts, e);
+            throw new ConditionTimeoutException("Failed to wait for test run " + runToWaitFor + " " + ts, e);
         }
     }
 
@@ -87,6 +89,8 @@ public class ContinuousTestingClient {
 
             return new TestStatus(lastRun, running, testsRun, testsPassed, testsFailed, testsSkipped, totalTestsPassed,
                     totalTestsFailed, totalTestsSkipped);
+        } catch (RuntimeException e) {
+            throw e;
         } catch (Exception e) {
             throw new RuntimeException(e);
         }

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/continuoustesting/ContinuousTestingLogClient.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/continuoustesting/ContinuousTestingLogClient.java
@@ -1,0 +1,156 @@
+package io.quarkus.gradle.continuoustesting;
+
+import static org.awaitility.Awaitility.await;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Utilities for testing continuous testing by parsing log output.
+ * <p>
+ * This is an alternative to {@link ContinuousTestingClient} that doesn't require
+ * an HTTP endpoint (DevUI). It parses the build output log for test results,
+ * similar to Maven's TestModeContinuousTestingMavenTestUtils.
+ * <p>
+ * Use this for quarkusTest mode where there's no HTTP server for the application.
+ */
+public class ContinuousTestingLogClient {
+
+    // Patterns to parse test results from the console output
+    // Example: "All 1 test is passing (0 skipped)"
+    // Example: "All 2 tests are passing (0 skipped)"
+    // Example: "1 test failed (0 passing, 0 skipped)"
+    // Example: "2 tests failed (1 passing, 0 skipped)"
+    // Note: The output may contain ANSI escape codes, so we strip them before matching.
+    private static final Pattern ALL_PASSING = Pattern.compile(
+            "All (\\d+) tests? (?:are|is) passing \\((\\d+) skipped\\)");
+    private static final Pattern SOME_FAILING = Pattern.compile(
+            "(\\d+) tests? failed \\((\\d+) passing, (\\d+) skipped\\)");
+    private static final String TESTS_COMPLETED = "Tests completed at";
+    // Pattern to strip ANSI escape codes
+    private static final Pattern ANSI_ESCAPE = Pattern.compile("\\x1B\\[[;\\d]*m");
+
+    private final File logFile;
+    private int logPosition = 0;
+
+    public ContinuousTestingLogClient(File logFile) {
+        this.logFile = logFile;
+    }
+
+    /**
+     * Waits for the next test run to complete and returns the results.
+     *
+     * @return the test status after completion
+     */
+    public TestStatus waitForNextCompletion() {
+        await()
+                .pollDelay(2, TimeUnit.SECONDS)
+                .atMost(5, TimeUnit.MINUTES)
+                .until(() -> {
+                    String log = getLogSinceLastCheck();
+                    boolean found = log.contains(TESTS_COMPLETED);
+                    if (!found && !log.isEmpty()) {
+                        // Debug: print a snippet of the log to help diagnose issues
+                        final boolean logFileExists = logFile.exists();
+                        System.out.println(
+                                "[ContinuousTestingLogClient] Waiting for tests... Log file exists: " + logFileExists
+                                        + ", size: " + (logFileExists ? logFile.length() : 0)
+                                        + ", checking from position: " + logPosition);
+                    }
+                    return found;
+                });
+
+        String log = getLogSinceLastCheck();
+        TestStatus result = parseTestResult(log);
+
+        // Update position for next check
+        updateLogPosition();
+
+        return result;
+    }
+
+    private String getLogSinceLastCheck() {
+        try {
+            if (!logFile.exists()) {
+                return "";
+            }
+            String fullLog = Files.readString(logFile.toPath());
+            return fullLog.substring(Math.min(logPosition, fullLog.length()));
+        } catch (IOException e) {
+            return "";
+        }
+    }
+
+    private void updateLogPosition() {
+        try {
+            if (logFile.exists()) {
+                logPosition = Files.readString(logFile.toPath()).length();
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to read log file", e);
+        }
+    }
+
+    private TestStatus parseTestResult(String log) {
+        // Strip ANSI escape codes before matching
+        String cleanLog = ANSI_ESCAPE.matcher(log).replaceAll("");
+
+        Matcher matcher = ALL_PASSING.matcher(cleanLog);
+        if (matcher.find()) {
+            int passed = Integer.parseInt(matcher.group(1));
+            int skipped = Integer.parseInt(matcher.group(2));
+            return new TestStatus(passed, 0, skipped);
+        }
+
+        matcher = SOME_FAILING.matcher(cleanLog);
+        if (matcher.find()) {
+            int failed = Integer.parseInt(matcher.group(1));
+            int passed = Integer.parseInt(matcher.group(2));
+            int skipped = Integer.parseInt(matcher.group(3));
+            return new TestStatus(passed, failed, skipped);
+        }
+
+        throw new AssertionError("Could not parse test results from log. " +
+                "Expected pattern like 'All N tests are passing' or 'N tests failed'. Log content:\n" + cleanLog);
+    }
+
+    /**
+     * Test result status.
+     */
+    public static class TestStatus {
+        private final int testsPassed;
+        private final int testsFailed;
+        private final int testsSkipped;
+
+        public TestStatus(int testsPassed, int testsFailed, int testsSkipped) {
+            this.testsPassed = testsPassed;
+            this.testsFailed = testsFailed;
+            this.testsSkipped = testsSkipped;
+        }
+
+        public int getTestsPassed() {
+            return testsPassed;
+        }
+
+        public int getTestsFailed() {
+            return testsFailed;
+        }
+
+        public int getTestsSkipped() {
+            return testsSkipped;
+        }
+
+        @Override
+        public String toString() {
+            return "TestStatus{" +
+                    "testsPassed=" + testsPassed +
+                    ", testsFailed=" + testsFailed +
+                    ", testsSkipped=" + testsSkipped +
+                    '}';
+        }
+    }
+}

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/CompileOnlyContinuousTestingModeTest.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/CompileOnlyContinuousTestingModeTest.java
@@ -1,0 +1,67 @@
+package io.quarkus.gradle.devmode;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.io.File;
+
+import com.google.common.collect.ImmutableMap;
+
+import io.quarkus.gradle.continuoustesting.ContinuousTestingLogClient;
+
+/**
+ * Tests that compile-only dependencies (like Lombok) are available to the compiler
+ * during live reload in quarkusTest mode.
+ */
+public class CompileOnlyContinuousTestingModeTest extends QuarkusDevGradleTestBase {
+
+    @Override
+    protected String projectDirectoryName() {
+        return "compile-only-lombok";
+    }
+
+    @Override
+    protected String[] buildArguments() {
+        return new String[] { "quarkusTest" };
+    }
+
+    @Override
+    protected void testDevMode() throws Exception {
+        // In quarkusTest mode, there's no persistent HTTP endpoint.
+        // Use the log-based client to parse test results from the build output.
+        File logFile = new File(getProjectDir(), "command-output.log");
+        ContinuousTestingLogClient testingClient = new ContinuousTestingLogClient(logFile);
+
+        // Wait for the initial test run to complete
+        ContinuousTestingLogClient.TestStatus results = testingClient.waitForNextCompletion();
+        assertEquals(1, results.getTestsPassed(),
+                "Initial test run should pass, actual results: " + results);
+        assertEquals(0, results.getTestsFailed(),
+                "Initial test run should have no failures, actual results: " + results);
+
+        // Modify source files that use Lombok annotations.
+        // This triggers live reload which requires the Lombok compiler to process annotations.
+        // Before the fix for issue #50936, this would fail with
+        // "package lombok.extern.slf4j does not exist" in quarkusTest mode.
+        replace("src/main/java/io/playground/MyData.java",
+                ImmutableMap.of("private String other;", "private final String other;"));
+        replace("src/main/java/io/playground/MyDataProducer.java",
+                ImmutableMap.of("return new MyData(\"lombok\");", "return new MyData(\"lombok\", \"!\");"));
+        replace("src/main/java/io/playground/HelloResource.java",
+                ImmutableMap.of("return \"hello \" + myData.getMessage();",
+                        "return \"hello \" + myData.getMessage() + myData.getOther();"));
+
+        // Update the test to expect the new response
+        replace("src/test/java/io/playground/HelloResourceTest.java",
+                ImmutableMap.of(".body(is(\"hello lombok\"));",
+                        ".body(is(\"hello lombok!\"));"));
+
+        // Wait for the next test run - this verifies that compile-only dependencies (Lombok)
+        // are available to the compiler during live reload.
+        // Before the fix, this would fail with "package lombok.extern.slf4j does not exist".
+        results = testingClient.waitForNextCompletion();
+        assertEquals(1, results.getTestsPassed(),
+                "Test run after live reload should pass, actual results: " + results);
+        assertEquals(0, results.getTestsFailed(),
+                "Test run after live reload should have no failures, actual results: " + results);
+    }
+}

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/QuarkusDevGradleTestBase.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/QuarkusDevGradleTestBase.java
@@ -7,6 +7,7 @@ import java.io.BufferedReader;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
@@ -20,10 +21,10 @@ import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-import org.apache.commons.io.FileUtils;
 import org.junit.jupiter.api.Test;
 import org.wildfly.common.Assert;
 
+import io.quarkus.bootstrap.util.IoUtils;
 import io.quarkus.gradle.BuildResult;
 import io.quarkus.gradle.QuarkusGradleWrapperTestBase;
 import io.quarkus.test.devmode.util.DevModeClient;
@@ -33,7 +34,11 @@ public abstract class QuarkusDevGradleTestBase extends QuarkusGradleWrapperTestB
     private Future<?> quarkusDev;
     protected File projectDir;
 
-    protected DevModeClient devModeClient = new DevModeClient();
+    protected int getPort() {
+        return 8080;
+    }
+
+    protected DevModeClient devModeClient = new DevModeClient(getPort());
 
     @Override
     protected void setupTestCommand() {
@@ -106,7 +111,7 @@ public abstract class QuarkusDevGradleTestBase extends QuarkusGradleWrapperTestB
             throw e;
         } finally {
             if (projectDir != null && projectDir.isDirectory()) {
-                FileUtils.deleteQuietly(projectDir);
+                IoUtils.recursiveDelete(projectDir.toPath());
             }
         }
     }
@@ -134,9 +139,9 @@ public abstract class QuarkusDevGradleTestBase extends QuarkusGradleWrapperTestB
         if (projectDir == null) {
             final String projectDirName = projectDirectoryName();
             try {
-                final File projectDir = Files.createTempDirectory(projectDirName).toFile();
-                FileUtils.copyDirectory(getProjectDir(projectDirName), projectDir);
-                this.projectDir = projectDir;
+                final Path projectDir = Files.createTempDirectory(projectDirName);
+                IoUtils.copy(getProjectDir(projectDirName).toPath(), projectDir);
+                this.projectDir = projectDir.toFile();
             } catch (Exception e) {
                 throw new IllegalStateException("Failed to create a project dir for " + projectDirName, e);
             }

